text/0000-safecoin-alternative-implementation/0000-safecoin-alternative-implementation.md

```diff
@@ -0,0 +1,238 @@
+
+On receipt of a `CoinTransfer` request (after passing out of Parsec), the CoinManagers will deduct the amount specified in `credit` from the source account.  If the account's balance doesn't permit this, the request will be silently dropped.
+
+These source CoinManagers will then send the `Credit` to the destination CoinManagers, where it will be run through Parsec to cover the case where that owner is concurrently trying to spend coin from that `CoinAccount`.
+
+If the account to be credited doesn't exist, the destination CoinManagers will recycle the coin by deducting the value from their section's `farmed` value.  The client will receive no notification that the transaction failed.  (We could look to refund the source `CoinAccount` in such cases, but this would require more effort by the network.  Well-designed client applications should be able to reduce the risk of accidental loss of coins in this way to zero.)
+
+Each `CoinAccount` will have an associated fixed-length FIFO queue for holding the most recent `transaction_id`s and `amount`s of `Credit`s made to that account.  When crediting the destination `CoinAccount`, the CoinManagers will push the `transaction_id` and `amount` onto that queue if the `transaction_id` is not `None`.  At this stage, the transaction is complete.
+
+Any client will be able to query for the existence of such a completed transaction by sending a `GetTransaction` message to the destination `CoinAccount`'s section (the destination address defining the particular `CoinAccount::id`) and expecting to receive a `Transaction` in response:
+
+```rust
+struct GetTransaction {
+    transaction_id: u64,
+}
+
+enum Transaction {
+    Success {
+        value: Coin,
+        transaction_id: u64,
+    },
+    NoSuchTransaction,
+    NoSuchCoinAccount,
+}
+```
+
+A CoinManager receiving a `GetTransaction` will always return a `Transaction`.  If the specified `CoinAccount` doesn't exist, `NoSuchCoinAccount` is returned.  If the `CoinAccount` does exist, but the requested `transaction_id` doesn't exist in its FIFO queue, `NoSuchTransaction` is returned.  Otherwise the appropriate `Success` value is returned.  As this message is signed by the Vaults comprising the CoinManagers, a `Transaction::Success` can be used as proof by any client, (in particular the sender) that the transaction was successful.
+
+This has the benefit of requiring neither the sender nor the receiver to stay connected while the payment is being processed, as would be the case for example if the sender needed to receive a response in order to prove payment had been made.  It also serves to further anonymise the sender, as the recipient will likely not be aware of the actual `CoinAccount` used by the sender to credit its own; only the `transaction_id` is visible.
+
+
+
+## Drawbacks
+
+A potential drawback of this proposal is that coins which are recycled aren't replaced back to the section where they were initially farmed.
+
```

Comment:
I see that a feature and a good one :-)

Reply:
Yeah, I'm not convinced this is a bad thing either, but it was a point Qi made when we were discussing my approach recently 😄  I think you're right; I'll probably remove this and let Qi make the case himself.